Samples/BackgroundBlur/BackgroundBlur/TransformAsync.cpp

```diff
@@ -0,0 +1,1212 @@
+#include "TransformAsync.h"
+#include <winrt/windows.media.h>
+#include <winrt/windows.foundation.h>
+#include <windows.graphics.directx.direct3d11.interop.h>
+#include <winrt/Windows.Graphics.DirectX.Direct3D11.h>
+#include <windows.graphics.imaging.interop.h>
+#include <windows.media.core.interop.h>
+#include "Helpers/common.h"
+#include <unknwn.h>
+#include <stdio.h>
+#include <wchar.h>
+#include <future>
+
+#define MFT_NUM_DEFAULT_ATTRIBUTES  4
+using namespace MainWindow;
+
+long long g_now; // The time since the last call to FrameThreadProc
+
+// Video FOURCC codes.
+const FOURCC FOURCC_NV12 = MAKEFOURCC('N', 'V', '1', '2');
+const FOURCC FOURCC_RGB24 = 20;
+const FOURCC FOURCC_RGB32 = 22;
+float runningAverage = 0.0f;
+
+// static array of media types (preferred and accepted).
+const GUID* g_MediaSubtypes[] =
+{
+    //&MFVideoFormat_NV12,
+    &MFVideoFormat_RGB32
+};
+
+// number of media types in the aray.
+DWORD g_cNumSubtypes = ARRAY_SIZE(g_MediaSubtypes);
+HRESULT GetImageSize(FOURCC fcc, UINT32 width, UINT32 height, DWORD* pcbImage);
+
+HRESULT TransformAsync::CreateInstance(IMFTransform** ppMFT)
+{
+    HRESULT hr = S_OK; 
+    com_ptr<TransformAsync> pMFT;
+    if (ppMFT == NULL)
+    {
+        return E_POINTER;
+    }
+    pMFT.attach(new TransformAsync(hr));
+    if (pMFT == NULL)
+    {
+        CHECK_HR(hr = E_OUTOFMEMORY);
+    }
+
+    CHECK_HR(hr = pMFT->InitializeTransform());
+    CHECK_HR(hr = pMFT->QueryInterface(IID_IMFTransform, (void**)ppMFT))
+
+done: 
+    return hr;
+}
+
+TransformAsync::TransformAsync(HRESULT& hr) 
+{
+    m_nRefCount = 1;
+    m_ulSampleCounter = 0;
+    m_spInputType = NULL;
+    m_spOutputType = NULL;
+    m_spAttributes = NULL;
+    m_spEventQueue = NULL;
+    m_dwStatus = 0;
+    m_dwNeedInputCount = 0;
+    m_dwHaveOutputCount = 0;
+    m_bShutdown = FALSE;
+    m_bFirstSample = TRUE;
+    // m_bDXVA = TRUE; // TODO: Always true to be fast
+    m_pInputSampleQueue = NULL;
+    m_pOutputSampleQueue = NULL;
+    m_videoFOURCC = 0;
+    m_imageWidthInPixels = 0; 
+    m_imageHeightInPixels = 0;
+
+    m_spDeviceManager = NULL;
```

Comment:
I would use the pattern to init all variables in header where they are declared 

example 

volatile ULONG                  m_nRefCount; = 1;